### ISSUE TITLE
Feature idea: Operate on text streams instead of files (i.e. stdin)

### ISSUE THREAD

**Issue by qwertystop** [jcjohnson/torch-rnn#172]
As best as I can tell, the main potential obstacle is the preprocessing. Allowing the necessary information to be given explicitly instead of by scanning the input file (or, perhaps, to be built by scanning only a sample of input) would allow e.g. piping in a stream from a network connection or some other external device. Perhaps pipe in streamed music through a processor that converts it to a stream of uncompressed 32-bit samples (if I'm not misreading the preprocessor script, you do support `uint32` as the token data type), and learn music directly from Pandora or YouTube. Or just take in ridiculously large bodies of ordinary text; it's all bytes in the end anyway.

**Comment by ChrisCummins** [comment]
One potential problem you could run into is determining the vocabulary to use. Currently, the vocabulary is determined by the characters which are in the corpus. If you don't know the corpus ahead of time (because you're streaming input), this becomes harder. You would either have to silently ignore out-of-vocabulary tokens, crash, or use (for character model) the entire possible ASCII/UTF inputs.

**Comment by qwertystop** [comment]
True. I would argue that's not much of a problem, though. If your input is arbitrary text, then yes, you want the full character set or some known subset. If it's unencoded, you want the full range of one-byte values. If it's formatted in some way, presumably the format is well-defined and any invalid character is a problem.

**Comment by antihutka** [comment]
It shouldn't be too hard to expand the vocabulary of the model during training. You'd have to resize the weight and bias tensors in the final Linear layer, copy the old values over and find a way to initialize the new values that works well enough.

**Comment by ChrisCummins** [comment]
I suspect that training times when using the full character set as opposed to only the subset which is required. @antihutka 's suggestion is nice. Whatever the approach, I'd love to see an implementation, as iterative training would be useful for a whole bunch of situations.

**Comment by qwertystop** [comment]
I don't know enough about the internals for @antihutka's suggestion, and the simple case that I do know how to do ("allow the preprocessor to operate on file-like streams") is probably less than helpful without the aforementioned knowledge of internals necessary to adapt the Torch code to also operate on streams.
